keps/sig-api-machinery/20190422-cloud-controller-manager-migration.md

```diff
@@ -0,0 +1,281 @@
+kube-controller-manager and the cloud-controller-manager via a shared resource lock between the two components. The pattern proposed
+in this KEP should be reusable by other components in the future if desired.
+
+### Goals
+
+* Define a clear migration process for existing clusters to use the cloud-controller-manager instead of the kube-controller-manager for cluster integrations with any cloud provider.
+* The migration process should work for single node control planes and HA control planes with minimal risk to users.
+
+### Non-Goals
+
+* Removing cloud provider code from the core Kubernetes tree, this effort is separate and is covered in [KEP-removing-in-tree-providers](https://github.com/kubernetes/enhancements/blob/master/keps/sig-cloud-provider/20190125-removing-in-tree-providers.md)
+* Improving the scalability of controllers by running controllers across multiple components (with or without leader election).
+* Migrating cloud-based volume plugins to CSI. This is a separate effort led by SIG Storage. See [this proposal](https://github.com/kubernetes/community/blob/master/contributors/design-proposals/storage/csi-migration.md) for more details.
+
+## Proposal
+
+Today, the kube-controller-manager and cloud-controller-manager run independent of each other.
+This means that both the kube-controller-manager or the cloud-controller-manager can run the
+cloud specific control loops for a given cluster. For a highly available control plane to migrate
+from running only the kube-controller-manager to running both the kube-controller-manager and
+the cloud-controller-manager requires that only one process in the control plane is running
+the cloud specific controllers. This becomes non-trivial when introducing the cloud-controller-manager
+that runs overlapping controllers with the kube-controller-manager.
+
+For environments that can tolerate downtime, it will be recommended to shutdown the control plane,
+reconfigure components to use the cloud-controller-manager, and then bring the control plane back up.
+This ensures that only 1 component can be running the set of cloud controllers. For environments
+that have stricter requirements for uptime, some level of coordination is required between the two
+components to ensure that upgrading control planes does not result in running the same controller
+in more than one place while also accounting for version skew.
+
+In order to coordinate the cloud-specific controllers across the `kube-controller-manager` and the `cloud-controller-manager`,
+this KEP proposes a _primary_ and _secondary_ (a.k.a migration) leader election lock in the `kube-controller-manager`. The primary lock represents
+the current leader election resource lock in the `kube-controller-manager`. The secondary lock would hold the run callback for all
+cloud-specific controllers in the `kube-controller-manager`. The secondary lock would only be run inside the run callback of the primary
```

Comment:
I think its worth calling out a level of indirection (generic) here. I think we want the secondary lock on a set of configurable controllers. Then we can configure that set of controllers to be the cloud specific controllers. We know that this set of controllers will be different for different cloud providers.